File: 백준/Silver/9012. 괄호/괄호.py
import sys
line = int(sys.stdin.readline())
stack=[]

for i in range(line):
    alist=list(sys.stdin.readline())
    for i in range(len(alist)-1):
        if alist[0]=="(":
            for u in range(1,len(alist)):
                if alist[u]==")":
                    alist.remove("(")
                    alist.remove(")")
                    break
        else:
            break
    if len(alist)==1 and alist[0]=="\n":
        print("YES")
    else:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.java

    import java.io.IOException;
    import java.io.BufferedReader;
    import java.io.InputStreamReader;
    import java.util.ArrayList;
    import java.util.List;

    public class Main {
        public static void main(String[] args) throws IOException {

            BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
            int len = Integer.parseInt(reader.readLine());
            for (int i = 0; i < len; i++) {
                String line = reader.readLine();
                String[] parts = line.split("");
                List<String> arrayList = new ArrayList<>();
                for (String part : parts) {

                    if (part.equals("(")) {

                        arrayList.add(part);

                    } else if (part.equals(")") && arrayList.contains("(") && arrayList.size() > 0) {
                        arrayList.remove("(");
                    } else if (part.equals(")")) {
                        arrayList.add(part);
                        break;
                    }
                }
                if(arrayList.isEmpty()){
                    System.out.println("YES");
                }else{
                    System.out.println("NO");
                }
                }
    }

            }


